File: RetroGameGauntlet.Forms/ViewModels/OverviewViewModel.cs
using System.Windows.Input;
using Plugin.Share;
using Xamarin.Forms;
using System.Threading.Tasks;
using RetroGameGauntlet.Forms.Services;
using System.Diagnostics;
using System.Collections.Generic;
using System;
using Plugin.Share.Abstractions;

namespace RetroGameGauntlet.Forms.ViewModels
{
    public class OverviewViewModel : BaseViewModel
    {
        private string _title;
        public string Title
        {
            get { return _title; }
            private set
            {
                _title = value;
                RaisePropertyChanged(nameof(Title));
            }
        }

        public bool IsClipboardEnabled
        {
            get { return CrossShare.Current.SupportsClipboard; }
        }

        public ImageSource _logoImageSource;
        public ImageSource LogoImageSource
        {
            get { return _logoImageSource; }
            set
            {
                _logoImageSource = value;
                RaisePropertyChanged(nameof(LogoImageSource));
            }
        }

        private string _description;
        public string Description
        {
            get { return _description; }
            set
            {
                _description = value;
                RaisePropertyChanged(nameof(Description));
            }
        }

        private string _searchButtonText;
        public string SearchButtonText
        {
            get { return _searchButtonText; }
            set
            {
                _searchButtonText = value;
                RaisePropertyChanged(nameof(SearchButtonText));
            }
        }

        #region `Click` properties

        private ICommand _shareClickCommand;
        public ICommand ShareClickCommand
        {
            get
            {
                _shareClickCommand = _shareClickCommand
                    ?? new Command(() => CrossShare.Current.Share(new ShareMessage
                    {
                        Text = string.Format("My {0} game is {1}", _platformName, _gameName),
                        Title = _gameName
                    }));
                return _shareClickCommand;
            }
        }

        private ICommand _copyToClipboardClickCommand;
        public ICommand CopyToClipboardClickCommand
        {
            get
            {
                _copyToClipboardClickCommand = _copyToClipboardClickCommand
                    ?? new Command(() => CrossShare.Current.SetClipboardText(string.Format("My {0} game is {1}", _platformName, _gameName)));
                return _copyToClipboardClickCommand;
            }
        }

        private ICommand _searchClickCommand;
        public ICommand SearchClickCommand
        {
            get
            {
                _searchClickCommand = _searchClickCommand ?? new Command(() => { });
                return _searchClickCommand;
            }
        }

        private IEnumerable<WikipediaItemViewModel> _wikipediaItems;
        public IEnumerable<WikipediaItemViewModel> WikipediaItems
        {
            get { return _wikipediaItems; }
            set
            {
                _wikipediaItems = value;
                RaisePropertyChanged(nameof(WikipediaItems));
            }
        }

        #endregion

        public event EventHandler<EventArgs> Initialized;

        private string _gameName;
        private string _platformName;

        private readonly IPlatformLoaderService _loaderService;
        private readonly IImageSearchService _imageSearchService;
        private readonly IWikipediaSearchService _wikiSearchService;

        public OverviewViewModel(IPlatformLoaderService loaderService,
                                 IImageSearchService imageSearchService,
                                 IWikipediaSearchService wikiSearchService)
        {
            _loaderService = loaderService;
            _imageSearchService = imageSearchService;
            _wikiSearchService = wikiSearchService;

            Title = "Loading...";
            SearchButtonText = "Find more";
        }

        public Task InitAsync(PlatformItemViewModel targetPlatform)
        {
            return Task.Factory.StartNew(async () =>
            {
                if (targetPlatform == null)
                {
                    return;
                }
                var newGameName = _loaderService.GetRandomGame(targetPlatform.PlatformModel.FileName);
                _gameName = newGameName;
                _platformName = targetPlatform.Title;
                await InitGameAsync();
            });
        }

        public Task InitAsync(KeyValuePair<string, string> targetGame)
        {
            return Task.Factory.StartNew(async () =>
            {
                _gameName = targetGame.Key;
                _platformName = targetGame.Value;
                await InitGameAsync();
            });
        }

        public Task InitListItemsAsync()
        {
            return Task.Factory.StartNew(async () =>
            {
                WikipediaItems = await _wikiSearchService.GetItemsForQuery(_gameName);
            });
        }

        private Task InitGameAsync()
        {
            return Task.Factory.StartNew(async () =>
            {

                var image = await _imageSearchService.GetImageForGame(_gameName, _platformName);
                Debug.WriteLine("The Flickr link is " + image);
                LogoImageSource = image;

                Title = _gameName;
                Description = string.Format("Your {0} game", _platformName);
                SearchButtonText = string.Format("Search \"{0} {1}\"", _platformName, _gameName);

                Initialized?.Invoke(this, EventArgs.Empty);
            });
        }
    }
}

File: RetroGameGauntlet.PCL/ViewModels/OverviewViewModel.fs
namespace RetroGameGauntlet.PCL.ViewModels

open System
open Plugin.Share
open Plugin.Share.Abstractions
open RetroGameGauntlet.PCL.Models
open RetroGameGauntlet.PCL.Services
open System.Collections.Generic
open System.Diagnostics
open System.Windows.Input
open Xamarin.Forms

/// The "Overview" page view model.
type OverviewViewModel(loaderService: IPlatformLoaderService,
                       imageSearchService: IImageSearchService,
                       wikiSearchService: IWikipediaSearchService) =
    inherit ViewModelBase()
    let mutable selectedGameName: string = String.Empty
    let mutable selectedPlatformName: string = String.Empty
    
    let shareMessage() = new ShareMessage(Text = String.Format("My {0} game is {1}", selectedPlatformName, selectedGameName), Title = selectedGameName)
    let shareClickCommand: ICommand = new Command(fun () -> 
                                                    CrossShare.Current.Share(shareMessage()) 
                                                    |> ignore) :> ICommand

    let clipboardText() = String.Format("My {0} game is {1}", selectedPlatformName, selectedGameName)
    let copyToClipboardClickCommand: ICommand = new Command(fun () -> 
                                                                CrossShare.Current.SetClipboardText(clipboardText())
                                                                |> ignore ) :> ICommand

    let searchClickCommand: ICommand = new Command(fun () -> Debug.WriteLine("Search")) :> ICommand

    let mutable title: string = "Loading..."
    let mutable logoImageSource: ImageSource = null
    let mutable description: string = System.String.Empty
    let mutable searchButtonText: string = "Find more"
    let mutable wikipediaItems: IEnumerable<WikipediaItemViewModel> = Seq.empty

    let eventInitialized = new Event<_>()

    /// The page title.
    member this.Title
        with get() = title
        and set parameter =
            title <- parameter
            this.NotifyPropertyChanged <@ this.Title @>

    /// A member shows if the current platform (e.g. Android) supports clipboarding.
    member this.IsClipboardEnabled: bool = CrossShare.Current.SupportsClipboard

    /// The game logo image source.
    member this.LogoImageSource
        with get() = logoImageSource
        and set parameter =
            logoImageSource <- parameter
            this.NotifyPropertyChanged <@ this.LogoImageSource @>

    /// The game description.
    member this.Description
        with get() = description
        and set parameter =
            description <- parameter
            this.NotifyPropertyChanged <@ this.Description @>

    /// The "Search" button text.
    member this.SearchButtonText
        with get() = searchButtonText
        and set parameter =
            searchButtonText <- parameter
            this.NotifyPropertyChanged <@ this.Description @>

    /// The "Share game" button command.
    member this.ShareClickCommand = shareClickCommand

    /// The "Copy to clipboard" command.
    member this.CopyToClipboardClickCommand = copyToClipboardClickCommand

    /// The "Search" button command.
    member this.SearchClickCommand = searchClickCommand

    /// The "Wikipeadia" listview source.
    member this.WikipediaItems
        with get() = wikipediaItems
        and set parameter =
            wikipediaItems <- parameter
            this.NotifyPropertyChanged <@ this.Description @>
    
    /// An event fires when the ViewModel is initialized.
    [<CLIEvent>]
    member public this.Initialized = eventInitialized.Publish

    /// Inits the game info according to the selectd platform.
    member public this.InitAsync (targetPlatform: PlatformItemViewModel) : Async<unit> =
        async {
            Debug.WriteLine("OverviewViewModel: Initializing...")
            let! newGameName = loaderService.GetRandomGameAsync(targetPlatform.PlatformModel.FileName) |> Async.AwaitTask
            let gameName = newGameName
            let platformName = targetPlatform.Title
            return! this.InitGameAsync(gameName, platformName)
        }
    
    /// Inits the game info according to the selected game.
    member public this.InitAsync (targetGame: KeyValuePair<string, string>) : Async<unit> =
        async {
            Debug.WriteLine("OverviewViewModel: Initializing...")
            let gameName = targetGame.Key
            let platformName = targetGame.Value
            return! this.InitGameAsync(gameName, platformName)
        }

    /// Inits the Wikipedia listview source.
    member public this.InitWikipediaItemsAsync () : Async<unit> =
        async {
            let! items = wikiSearchService.GetItemsForQueryAsync selectedGameName
            this.WikipediaItems <- items
        }

    /// Inits the game info according to the selected platform and game.
    member private this.InitGameAsync (gameName: string, platformName: string) : Async<unit> =
        async {
            let! image = imageSearchService.GetImageForGameAsync gameName platformName
            Debug.WriteLine("The Flickr link is " + image)
            if String.IsNullOrEmpty image |> not then
                this.LogoImageSource <- UriImageSource(Uri=Uri(image))
            selectedGameName <- gameName
            selectedPlatformName <- platformName
            this.Title <- gameName
            this.Description <- String.Format("Your {0} game", platformName)
            this.SearchButtonText <- String.Format("Search for \"{0} {1}\"", platformName, gameName)
            eventInitialized.Trigger(this, EventArgs.Empty)
        }